### PR TITLE
cleaned up exception handling slightly

### LINKED ISSUE #50: error during exception handling: tried to access `cause` field of `ErrorException`
An error can occur during exception handling wherein the (non-existant) `cause` field of an `ErrorExcept` (or other error) can be accessed.  This occurs on line 446 of AWSCore.jl.

PR incoming.

### PR DESCRIPTION
**Pull request by ExpandingMan**
This fixes #50 as well as cleaning up the exception handling a bit.  So the issue is that `AWSException` had a constructor that was just the identity, and that this was being called on `ErrorException`.  It then got to a line where it tried to access a field of `ErrorException` and crashed during error handling.

I've done a couple of things here to try to improve the situation.  First, I've removed the identity constructor of `AWSException`.  This constructor seems very confusing since not only might it return an unexpected `Exception` type, it can in fact return literally anything at all.

I have replaced this constructor with a new `aws_exception` function (non necessarily the best name).  The idea is that this function should construct an `AWSException` on any exception which `AWSException` can wrap (at the moment, only `HTTP.StatusError`) but if passed an exception which it does *not* know how to handle (as was happening in #50) it will immediately rethrow the error.  This way the body of `do_request` is free to implement handling of `AWSException` knowing that all other exceptions will be safely thrown.  It's possible that more methods should be added to this, but there were none that I could see for right now (certainly more can be added in the future).

I also took the liberty of removing the `__precompile__` call which is not necessary in Julia 1.0.

I realize that this was a somewhat aggressive fix to this small issue, but I'll hope you'll agree with me that the identity constructor for `AWSException` was confusing and likely would have led to more errors in the future.  Thanks!

### CODE REVIEW

File: src/AWSException.jl [hunk 1]
```diff
@@ -56,7 +56,8 @@ function AWSException(e::HTTP.StatusError)
 end
 
 
-AWSException(e) = e
+aws_exception(e::Exception) = rethrow(e)
```
Comment (iamed2) [inline src/AWSException.jl]: Please add docstrings :)

**Comment by samoconnor** [comment]
Thanks for looking into this @ExpandingMan !
I found the same issue while deep in the detail of testing a fix for https://github.com/JuliaWeb/MbedTLS.jl/issues/174, but I haven't got to pushing the fix yet. Sorry.

I agree that `AWSException` == `identity` for some input made the code hard to follow.

Note that the other handling in that block uses `ecode(e)` which checks if `e` has a code field and returns `nothing` by default: https://github.com/samoconnor/Retry.jl/blob/master/src/Retry.jl#L7-L8

This is the patch that I had made but not commited yet for this issue:
I have this sitting uncommitted in my local copy:
```patch
--- a/src/AWSCore.jl
+++ b/src/AWSCore.jl
@@ -443,7 +443,8 @@ function do_request(r::AWSRequest)
         # Recommended for SDKs at:
         # https://docs.aws.amazon.com/general/latest/gr/api-retries.html
         # Also BadDigest error and CRC32 thing
-        @retry if http_status(e.cause) == 429 ||
+        @retry if e isa AWSException && (
+                  http_status(e.cause) == 429 ||
                   ecode(e) in ("Throttling",
                                "ThrottlingException",
                                "ThrottledException",
@@ -456,7 +457,7 @@ function do_request(r::AWSRequest)
                                "BadDigest",
                                "RequestTimeoutException",
                                "PriorRequestNotComplete") ||
-                  HTTP.header(e.cause.response, "crc32body") == "x-amz-crc32" end
+                  header(e.cause, "crc32body") == "x-amz-crc32") end
     end
```

In the context of a [`Retry.@repeat`](https://github.com/samoconnor/Retry.jl#repeat-n-try) block, `aws_exception(e::Exception) = rethrow(e)` doesn't  seem right because the rethrow is automatic.

I think the best way to make it clear to future readers that `e` is wrapped but only in some cases is to do this :
```patch
--- a/src/AWSCore.jl
+++ b/src/AWSCore.jl
@@ -421,7 +421,9 @@ function do_request(r::AWSRequest)
         end

     catch e
-        e = AWSException(e)
+        if e isa HTTP.StatusError
+            e = AWSException(e)
+        end
```
 ```patch
--- a/src/AWSException.jl
+++ b/src/AWSException.jl
@@ -56,9 +56,6 @@ function AWSException(e::HTTP.StatusError)
 end


-AWSException(e) = e
-
-
```

**Comment by samoconnor** [comment]
See: https://github.com/JuliaCloud/AWSCore.jl/commit/25b8e4b6f74ad51095fc5087e4d4930c45509da5

**Comment by ExpandingMan** [comment]
I see what you're saying about hiding the `rethrow` being undesirable.  On the other hand it seems like a shame to give up good multiple dispatch code in favor of checking everything, it seems to me like much of what's being checked in this file could be dispatched on instead, and this could be a starting point.

Anyway, your fix clearly solves the issue, so if you don't want to do it this way, of course feel free to close the PR.  Thanks for the fix.

**Comment by samoconnor** [comment]
> good multiple dispatch code

😅

Multiple dispatch is a good tool, but it's one of many tools in the toolbox.
If the generic meaning of the function is clear from the name, and if the method implementers are careful keep to the spirit of the generic meaning, then multiple dispatch is great. But there are other times when you can end up in `@which` or `grep` hell trying to figure out "what does this line of code do". I think especially when it comes to error state handling code the first priority is to be unambiguous (which my old `AWSException(e) = e` certainly was not!).

That being said, there are a few patterns in this code that are less than ideal. It was my first Julia code and it was written back in v0.3 days. So, please do make suggestions (or PRs) if you can see ways to improve it.

**Comment by ExpandingMan** [comment]
Sorry, that probably seemed silly or petty in the context of this one issue in isolation.  The idea is that handling objects in a general way that does not depend on implementation details in contexts which are general (such as the type of an exception in a `try` `catch` block) will avoid exactly these types of errors.  In my experience if you have to spend much time on `@which` or `grep` to figure out what is happening somethings not being done right, one shouldn't use multiple dispatch to hide the meaning of functions, methods of functions should have a similar and predictable behavior.  It becomes more difficult when there are no type assertions on function arguments anywhere, not only do (modest) type assertions catch strange errors, it tells the package developer (and the user) what functions do.

Anyway, calling `rethrow` through `aws_exception` probably fails to live up to this, as that's rather different than the context of `rethrow` in the rest of the block.  And, besides, much of what I'm saying is probably a moot point in this package anyway since we are dealing with a generic web API wrapper and IO where in many cases we'd struggle to determine types anyway!

I'll close, thanks again for the fix.
